README.md
script.module.neverwise
================

NeverWise XBMC toolkit (tested on XBMC 13.1 Gotham)

You may need to download the dependencies of this addon:
- script.module.chardet 2.1.2 (http://mirrors.xbmc.org/addons/frodo/script.module.chardet/script.module.chardet-2.1.2.zip)
- script.module.beautifulsoup 3.2.1 (http://mirrors.xbmc.org/addons/frodo/script.module.beautifulsoup/script.module.beautifulsoup-3.2.1.zip)
- script.module.parsedom 2.5.2 (http://mirrors.xbmc.org/addons/frodo/script.module.parsedom/script.module.parsedom-2.5.2.zip)


changelog.txt
[B]Version 1.1.0[/B]
- Updated BeautifulSoup to version 4.
- Added management of: cookies, datetime format, timezone and context menu.
- Added getDownloadContextMenu function.
- Updated user agent.
- Improved the getResponse functions.

[B]Version 1.0.9[/B]
- Added htmlDecode function.

[B]Version 1.0.8[/B]
- Improved the getResponse functions.
- Added showNotification function to replace dialog boxes with notifications.
- Removed timeout.
- Updated user agent.
- Added stripTags function (new version of old trimTags function).

[B]Version 1.0.7[/B]
- Setted 60 seconds of timeout.
- Updated user agent.

[B]Version 1.0.6[/B]
- Removed functions normalizeText and trimTags.
- createItemPage now is createNextPageItem.
- Added function createAudioVideoItems.
- getHtml function has been divided in getResponseJson, getResponseBS, getResponseForRegEx and getResponse.

[B]Version 1.0.5[/B]
- Updated formatUrl function.
- Added encodeDict function.
- Updated user agent.

[B]Version 1.0.4[/B]
- Small bugfix/improvements.

[B]Version 1.0.3[/B]
- Improved some functions.
- Adapted to the xbmc repository guidelines.

[B]Version 1.0.2[/B]
- Changed structure with the Util class.

[B]Version 1.0.1[/B]
- Added the BeautifulSoup library.
- Added error message on the Internet connection.
- Improved some functions.

[B]Version 1.0.0[/B]
- Initial version.


lib/tools.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Version 1.0.1 (27/12/2013)
# NeverWise XBMC toolkit
# Strumenti per i plugins di xbmc.
# By NeverWise
# <email>
# <web>
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#######################################################################
import re, sys, urllib, urllib2, urlparse, xbmcplugin, xbmcgui, xbmcaddon, BeautifulSoup, CommonFunctions


def getTranslation(addonId, translationId):
  return xbmcaddon.Addon(addonId).getLocalizedString(translationId).encode('utf-8')


def getBSResponseUrl(url):
  result = getResponseUrl(url)
  if result != None:
    result = BeautifulSoup.BeautifulSoup(result)
  return result


def getResponseUrl(url):
  html = None
  req = urllib2.Request(url, headers = { 'User-Agent' : 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:27.0) Gecko/20100101 Firefox/27.0' })

  try:
    response = urllib2.urlopen(req)
  except:
    xbmcgui.Dialog().ok('NeverWise toolkit', getTranslation('script.neverwise', 30001))
  else:
    html = response.read()
    response.close()

  if html != None:
    html = html.replace('\t', '').replace('\r\n', '').replace('\n', '').replace('\r', '').replace('" />', '"/>')
    while html.find('  ') > -1: html = html.replace('  ', ' ')

  return html


def normalizeText(text):
  text = text.decode('utf8', 'xmlcharrefreplace')
  return CommonFunctions.replaceHTMLCodes(text).strip()


def stripTags(html): return re.sub('<.+?>', '', html)


# Convert parameters encoded in a URL to a dict.
def urlParametersToDict(parameters):
  if len(parameters) > 0 and parameters[0] == '?':
    parameters = parameters[1:]
  return dict(urlparse.parse_qsl(parameters))


def createListItem(name, thumbimage, fanart, streamtype, infolabels):
  li = xbmcgui.ListItem(name, iconImage = 'DefaultFolder.png', thumbnailImage = thumbimage)
  li.setProperty('fanart_image', fanart)
  li.setInfo(streamtype, infolabels)
  return li


def addItem(handle, name, thumbimage, fanart, streamtype, infolabels, url, isFolder):
  li = createListItem(name, thumbimage, fanart, streamtype, infolabels)
  return xbmcplugin.addDirectoryItem(handle = handle, url = url, listitem = li, isFolder = isFolder)


def addLink(handle, name, thumbimage, fanart, streamtype, infolabels, url):
  return addItem(handle, name, thumbimage, fanart, streamtype, infolabels, url, False)


def addDir(handle, name, thumbimage, fanart, streamtype, infolabels, params):
  u = sys.argv[0] + '?' + urllib.urlencode(params)
  return addItem(handle, name, thumbimage, fanart, streamtype, infolabels, u, True)


lib/neverwise.py
#!/usr/bin/python
import cookielib, datetime, gzip, HTMLParser, json, os, re, StringIO, sys, urllib, urllib2, urlparse, xbmc, xbmcaddon, xbmcgui, xbmcplugin
import time # Workaround bug.
from bs4 import BeautifulSoup
from dateutil import tz

_idPlugin = 'script.module.neverwise'
addon = xbmcaddon.Addon()
addonName = addon.getAddonInfo('name')
icon_path = os.path.join(addon.getAddonInfo('path'), 'icon.png')
date_format = xbmc.getRegion('dateshort')
time_format = xbmc.getRegion('time')
datetime_format = '{0} {1}'.format(date_format, time_format)


def getResponseJson(url, headers = {}, show_error_msg = True):

  result = getResponse(url, headers, show_error_msg)

  if result.isSucceeded and len(result.body) > 0:
    try:
      result.body = json.loads(result.body)
    except:
      result.isSucceeded = False
      if show_error_msg:
        showResponseError()
  else:
    result.isSucceeded = False

  return result


def getResponseBS(url, headers = {}, show_error_msg = True, parser = 'html.parser'):

  result = getResponse(url, headers, show_error_msg)

  if result.isSucceeded:
    result.body = result.body.replace('&nbsp;', ' ')
    result.body = normalizeResponse(result.body)
    result.body = BeautifulSoup(result.body, parser)

  return result


def getResponseForRegEx(url, headers = {}, show_error_msg = True):

  result = getResponse(url, headers, show_error_msg)

  if result.isSucceeded:
    result.body = htmlDecode(result.body)
    result.body = normalizeResponse(result.body)

  return result


def getResponse(url, headers = {}, show_error_msg = True):

  defaultHeaders = {
    'User-Agent' : 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:53.0) Gecko/20100101 Firefox/53.0',
    'Accept-Encoding' : 'gzip, deflate'
  }

  for key, value in defaultHeaders.iteritems():
    if key not in headers:
      headers[key] = value

  cookies = cookielib.CookieJar()
  opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(cookies))
  urllib2.install_opener(opener)

  req = urllib2.Request(url, headers = headers)

  result = Response()
  try:
    response = urllib2.urlopen(req)
    result.cookies = cookies
  except:
    result.isSucceeded = False
    if show_error_msg:
      showConnectionError()
  else:
    result.body = response.read()
    encoding = response.info().get('Content-Encoding')
    if encoding == 'gzip':
      result.body = gzip.GzipFile(fileobj = StringIO.StringIO(result.body)).read()
    elif encoding == 'deflate':
      result.body = zlib.decompress(result.body)

    charset = response.headers.getparam('charset')
    if charset != None:
      result.body = result.body.decode(charset)

    response.close()

    if result.body.find(b'\0') > -1: # null bytes, if there's, the response is wrong.
      result.isSucceeded = False
      if show_error_msg:
        showResponseError()

  return result


def normalizeResponse(body):
  body = re.sub('\s+', ' ', body)
  return body.replace('" />', '"/>')


def showNotification(message, time = 5000, header = addonName, image = icon_path):
  notify = u'XBMC.Notification({0},{1},{2},{3})'.format(header, message, time, image)
  xbmc.executebuiltin(notify.encode('utf-8'))


def showConnectionError():
  showNotification(getTranslation(33001, _idPlugin))


def showVideoNotAvailable():
  showNotification(getTranslation(33002, _idPlugin))


def showResponseError():
  showNotification(getTranslation(33003, _idPlugin))


def getTranslation(translationId, addonId = ''):
  return xbmcaddon.Addon(addonId).getLocalizedString(translationId)


# Convert parameters encoded in a URL to a dict.
def urlParametersToDict(parameters):
  if len(parameters) > 0 and parameters[0] == '?':
    parameters = parameters[1:]
  return dict(urlparse.parse_qsl(parameters))


def createListItem(label, label2 = '', iconImage = None, thumbnailImage = None, path = None, fanart = None, streamtype = None, infolabels = None, duration = '', isPlayable = False, contextMenu = None):
  li = xbmcgui.ListItem(label, label2)

  if iconImage:
    li.setIconImage(iconImage)

  if thumbnailImage:
    li.setThumbnailImage(thumbnailImage)

  if path:
    li.setPath(path)

  if fanart:
    li.setProperty('fanart_image', fanart)

  if streamtype:
    li.setInfo(streamtype, infolabels)

  if streamtype == 'video' and duration:
    li.addStreamInfo(streamtype, {'duration': duration})

  if isPlayable:
    li.setProperty('IsPlayable', 'true')

  if contextMenu and len(contextMenu) > 0:
    li.addContextMenuItems(contextMenu)

  return li


def formatUrl(parameters, domain = sys.argv[0]):
  return '{0}?{1}'.format(domain, urllib.urlencode(encodeDict(parameters)))


def createNextPageItem(handle, pageNum, urlDictParams, fanart = None):
  title = '{0} {1} >'.format(getTranslation(33000, _idPlugin), pageNum)
  xbmcplugin.addDirectoryItem(handle, formatUrl(urlDictParams), createListItem(title, thumbnailImage = 'DefaultVideoPlaylists.png', fanart = fanart), True)


def createAudioVideoItems(handle, fanart = None):
  xbmcplugin.addDirectoryItem(handle, formatUrl({ 'content_type' : 'video' }), createListItem(getTranslation(33004, _idPlugin), thumbnailImage = 'DefaultMovies.png', fanart = fanart), True)
  xbmcplugin.addDirectoryItem(handle, formatUrl({ 'content_type' : 'audio' }), createListItem(getTranslation(33005, _idPlugin), thumbnailImage = 'DefaultMusicSongs.png', fanart = fanart), True)


def encodeDict(oldDict):
  newDict = {}
  for k, v in oldDict.iteritems():
    if isinstance(v, unicode):
      v = v.encode('utf8')
    elif isinstance(v, str):
      # Must be encoded in UTF-8
      v.decode('utf8')
    newDict[k] = v
  return newDict


def playStream(handle, label, thumbnailImage = None, path = None, streamtype = None, infolabels = None):
  li = createListItem(label, thumbnailImage = thumbnailImage, path = path, streamtype = streamtype, infolabels = infolabels)
  xbmcplugin.setResolvedUrl(handle, True, li)


def stripTags(text):
  text = re.sub('<[^<]+?>', '', text)
  return text.strip()


def htmlDecode(text):
  try:
    text = text.decode('utf-8')
  except:
    pass
  return HTMLParser.HTMLParser().unescape(text)


def getDownloadContextMenu(action, text = None):
  default_text = getTranslation(33006, _idPlugin)
  if text and len(text) > 0:
    return [(u'{0} {1}'.format(default_text, text), action)]
  else:
    return [(default_text, action)]


def strptime(date_string, date_format_string):
  try:
    return datetime.datetime.strptime(date_string, date_format_string)
  except TypeError:
    return datetime.datetime(*(time.strptime(date_string, date_format_string)[0:6])) # Workaround bug.


def gettz(name):
  return tz.gettz(name)


def gettzlocal():
  tzKodi = json.loads(xbmc.executeJSONRPC('{"jsonrpc": "2.0", "method": "Settings.GetSettingValue", "params": {"setting": "locale.timezone"}, "id": 1}'))
  if not 'error' in tzKodi:
    return tz.gettz(tzKodi['result']['value'])
  else:
    return tz.tzlocal()


class Response(object):
  body = None
  cookies = None
  isSucceeded = True
